rfd/0131-adminitrative-actions-mfa.md

```diff
@@ -0,0 +1,274 @@
+  - This can potentially be avoided within the implementation by introducing
+    a common gRPC message for the `MFAChallengeResponse`.
+
+1. Allow clients to pass `MFAChallengeResponse` as client request metadata.
+
+This approach is similar to passing a bearer-token in HTTP requests. Note that it
+will augment the normal certificate auth flow, not replace it. Additionally, the
+MFA Challenge Response will be passed to the server within the context of the TLS
+handshake, so it is secure.
+
+The client can cleanly pass the challenge response with the gRPC client call
+option `PerRPCCredentials`. The Auth server can then grab the `MFAChallengeResponse`
+from the request metadata, consume/verify the challenge response, and treat the request
+as MFA verified. [POC](https://github.com/gravitational/teleport/commit/aa0a8102eccd91cff2851053fba3e1c271bdaa65).
+
+Pros:
+
+- Client can apply common logic for MFA verification, rather than endpoint
+  specific logic.
+  - Ex: First request without MFA, receive access denied error, re-request with MFA.
+  - MFA verification can be extended to additional Auth endpoint without any client changes.
+- Adds a new system of passing user credentials, which can be used in other
+  endpoints that always or sometimes need MFA verification.
+  - Ex: `GenerateUserSingleUseCerts` could be replaced by simply checking for
+    MFA verification in the request metadata of `GenerateUserCerts`.
+- The MFA verification only applies to one API request, preventing MFA-replay-like
+  attacks.
+
+Cons:
+
+- Adding a new system of passing user credentials makes the client harder to
+  reason about, especially for external users of the API client.
+
+1. Reissue short-lived (1 minute TTL) MFA verified TLS certificates.
+
+This approach is similar to Per-session MFA. The client will need to issue a
+`GenerateUserSingleUseCerts` request to get MFA verified TLS certificates before
+making an administrative request.
+
+Pros:
+
+- Client can apply common logic for MFA verification, rather than endpoint
+  specific logic.
+  - Ex: First request without MFA, receive access denied error, re-request with MFA.
+  - MFA verification can be extended to additional Auth endpoint without any client changes.
+- Easily extends to other endpoints that always or sometimes need MFA verification.
+- Auth is completely handled by certificates rather than an augmented client request.
+
+Cons:
+
+- MFA verification applies to any requests using the MFA verified certificates.
+  This can be used as a MFA-replay-like attack since only one MFA verification is
+  required to carry out multiple administrative actions.
+  - Like Per-session MFA, this puts the security of the system more in the hands
+    of the client to not misuse or expose the certificates within their one
+    minute lifetime.
+- Requires users to create a new client with the reissued certificates, resulting
+  in additional connections to the Auth server.
+
+##### Option choice: TBD
```

Comment:
IMO option "3" is the most consistent with existing per-session MFA implementation. Others seems like inventing new auth protocols.

Reply:
I agree with @r0mant, and in general prefer option 3, assuming that the action is encoded into the cert as described above and that the expiration is extremely short (1 min?).

If we feel that we don't want to increase the scope, or that's such a TTL wont be acceptable, then the replay attack risk seems very significant to me.   Under those conditions option 1 or 2 seem like a better option to me.

Reply:
@jentfoo We were planning to introduce limited-scope certificates as part of the headless work but it got pushed out of scope. We still want to do it, but would you be okay with us pushing it out of initial scope of this feature and implementing later (say, Q3)? The MFA certificate do have 1 min TTL by design.